### PR TITLE
webgl: use a global gl context (no more limit of 16 webgl plots)

### PR DESCRIPTION
**Pull request by almarklein**
issues: #2590 

All plots in one page share the same gl context.

### CODE REVIEW

File: bokehjs/src/coffee/common/plot.coffee [hunk 1]
```diff
@@ -122,17 +122,17 @@ class PlotView extends ContinuumView
...
   init_webgl: () ->
 
-    # Get visible and invisible canvas 
-    # (doing it like this makes it easy to swap them during debugging)    
-    glcanvas = document.createElement('canvas')
-    
-    # Create GL context to draw to
+    # We use a global invisible canvas and gl context. By having a global context,
+    # we avoid the limitation of max 16 contexts that most browsers have. 
+    glcanvas = window._bokeh_gl_canvas
```
Comment (mattpap) [inline bokehjs/src/coffee/common/plot.coffee]: Does this have to be set on `window`? Maybe you could use a module-level variable?

Comment (almarklein) [inline bokehjs/src/coffee/common/plot.coffee]: Oh sure, as long as its "global" to all plots in a browser window. How would I use a module-level variable? Is there a global keyword in coffeescript?

Comment (mattpap) [inline bokehjs/src/coffee/common/plot.coffee]: No, but just initialize a variable at module-level and then you will be able to assign to it from inner scopes (google for "coffeescript scoping madness" for more details).

**Comment by almarklein** [comment]
@mattpap Oh wow, that is scary. Well, made use it to store the context at the module level.

**Comment by almarklein** [comment]
rebased on dev0.10, travis should pass now

File: bokehjs/src/coffee/common/plot.coffee [hunk 2]
```diff
@@ -328,7 +330,7 @@ class PlotView extends ContinuumView
       ctx.drawImage(ctx.glcanvas, 0.1, 0.1)
       for prefix in ['image', 'mozImage', 'webkitImage','msImage']
          ctx[prefix + 'SmoothingEnabled'] = true
-      console.log('drawing with WebGL')
+      #console.log('drawing with WebGL')
```
Comment (bryevdv) [inline bokehjs/src/coffee/common/plot.coffee]: We have an actual logger facility with log levels.

**Comment by damianavila** [comment]
>rebased on dev0.10, travis should pass now

@almarklein, please, do not rebase, merge instead... we know that in some other projects the rebase strategy is allowed after publishing the branch, but after some discussions, we agreed here to use a merge strategy after the branch is published. Just to let you know... we should probably make a note about that somewhere... Thanks!!

**Comment by almarklein** [comment]
@damianavila ok :)

**Comment by almarklein** [comment]
@bryevdv fixed to use logger.debug()

**Comment by almarklein** [comment]
ping

**Comment by bryevdv** [comment]
LGTM (have not actually checked out and tested)

**Comment by bryevdv** [comment]
Is there an easy way to test this, @almarklein ?

**Comment by almarklein** [comment]
Yes, the `plotting/file/glyphs.py` example fails when using WebGL, and works again with this PR. To test that it does not break anything you can test on any example with more than one plot in the page (with webgl turned on, of course).

**Comment by bryevdv** [comment]
OK I was specifically asking if there was an example with more than 16 webgl plots. :) I just duplicated the circle plot in `glyphs.py` around 20 times. LGTM (tested)

**Comment by almarklein** [comment]
Yes, `plotting/file/glyphs.py` has more than 16 plots, right? Well, what you did should do the trick as well :)

**Comment by bryevdv** [comment]
Oh, I guess I assumed that plots that can't/don't currently use any GL (all the non circle markers e.g.) would not contribute to the total.

**Comment by almarklein** [comment]
That depends. If you enable webgl by setting the plot's `webgl` property to True, then you're right. But when using `?webgl=1` in the address, all plots become "webgl-capable" and get a gl context (though if none of the glyphs inside a plot support webgl that context is never used).

**Pull request merged**